cloudwatch_metrics_collector/test/cloudwatch_metrics_collector_test.cpp

```diff
@@ -19,289 +19,393 @@
+#include <ros_monitoring_msgs/MetricData.h>
+
+#include <cloudwatch_metrics_collector/metrics_collector_parameter_helper.hpp>
+
+using namespace Aws::CloudWatchMetrics;
+using namespace Aws::CloudWatchMetrics::Utils;
+using ::testing::_;
+using ::testing::AllOf;
+using ::testing::HasSubstr;
+using ::testing::Return;
+using ::testing::StrEq;
+using ::testing::Eq;
+using ::testing::InSequence;
+
+int test_argc;
+char** test_argv;
+
+class MetricServiceFactoryMock : public MetricServiceFactory
+{
+public:
+    MOCK_METHOD4(createMetricService,
+            std::shared_ptr<MetricService>(
+            const std::string & metrics_namespace,
+            const Aws::Client::ClientConfiguration & client_config,
+            const Aws::SDKOptions & sdk_options,
+            const CloudWatchOptions & cloudwatch_option)
+    );
+};
+
+class MetricBatcherMock : public MetricBatcher
+{
+public:
+
+    MOCK_METHOD0(publishBatchedData, bool());
+};
+
+class MetricPublisherMock : public MetricPublisher
+{
+public:
+    MetricPublisherMock(const std::string & metrics_namespace,
+                        const Aws::Client::ClientConfiguration & client_config)
+            : MetricPublisher(metrics_namespace, client_config) {}
+};
+
+class MetricServiceMock : public MetricService
+{
+public:
+    MetricServiceMock(std::shared_ptr<Publisher<MetricDatumCollection>> publisher,
+                      std::shared_ptr<DataBatcher<MetricDatum>> batcher,
+                      std::shared_ptr<FileUploadStreamer<MetricDatumCollection>> file_upload_streamer = nullptr)
+            : MetricService(publisher, batcher, file_upload_streamer) {}
+
+    MOCK_METHOD1(batchData, bool(const MetricObject & data_to_batch));
+    MOCK_METHOD0(start, bool());
+    MOCK_METHOD0(shutdown, bool());
+    MOCK_METHOD0(publishBatchedData, bool());
+};
+
+
+class MetricsCollectorFixture : public ::testing::Test
+{
+protected:
+  const std::string kMetricsTopic = "metrics";
+  const std::string kMetricName1 = "CWMetricsNodeTestMetric";
+  const std::string kMetricUnit1 = "sec";
+  const std::string metric_namespace = "test_namespace";
+
+  Aws::Client::ClientConfiguration config;
+  Aws::SDKOptions sdk_options;
+  Aws::CloudWatchMetrics::CloudWatchOptions cloudwatch_options;
+
+  std::shared_ptr<MetricServiceMock> metric_service;
+  std::shared_ptr<MetricPublisherMock> metric_publisher;
+  std::shared_ptr<MetricBatcherMock> metric_batcher;
+  std::shared_ptr<MetricsCollector> metrics_collector;
+
+  std::shared_ptr<ros::NodeHandle> node_handle;
+  std::shared_ptr<ros::Publisher> metrics_pub;
+
+  void SetUp()
+  {
+    metric_batcher = std::make_shared<MetricBatcherMock>();
+    metric_publisher = std::make_shared<MetricPublisherMock>(metric_namespace, config);
+    metric_service = std::make_shared<MetricServiceMock>(metric_publisher, metric_batcher);
+  }
+
+  void Initialize(std::map<std::string, std::string> metric_dimensions) {
+
+    ros::init(test_argc, test_argv, "CWMetricsNodeTest");
```

Comment:
do we need to call ros::shutdown? or is that automatically done when this fixture is destroyed?

Reply:
> do we need to call ros::shutdown? or is that automatically done when this fixture is destroyed?

That's a good question, as I was lacking context for these tests and haven't received an answer yet. My possibly bad assumption was that roscore would already be running when these tests are executed.

Reply:
I think you're right. I think that's coming from the `test_cloudwatch_metrics_collector.test` file